From cc23959b02ea59b53d84f5561ceff1f6589fec63 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Wed, 19 Jun 2024 18:21:46 +0200
Subject: [PATCH 129/140] mailbox: add tx done by ack to all device having and
 fix mbox free

Devices having irq capability have the property tx_done_method set to
TXDONE_BY_IRQ, but they can support TXDONE_BY_ACK. Reinitialize tx_done method from channel at free channel.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/linux.git 28974436a2b431f5c80bb8b3ef40e6a2654e7f58 mailbox: add tx done by ack to all device having and fix mbox free]
Change-Id: I19d3ec92cf9d2048120566e579a0cd498a7b5a37
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/402234
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 drivers/mailbox/mailbox.c | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

diff --git a/drivers/mailbox/mailbox.c b/drivers/mailbox/mailbox.c
index 2acc6ec229a45..e624eed99fcb8 100644
--- a/drivers/mailbox/mailbox.c
+++ b/drivers/mailbox/mailbox.c
@@ -323,7 +323,7 @@ static int __mbox_bind_client(struct mbox_chan *chan, struct mbox_client *cl)
 		chan->cl = cl;
 		init_completion(&chan->tx_complete);
 
-		if (chan->txdone_method	== TXDONE_BY_POLL && cl->knows_txdone)
+		if (cl->knows_txdone)
 			chan->txdone_method = TXDONE_BY_ACK;
 	}
 
@@ -461,6 +461,16 @@ struct mbox_chan *mbox_request_channel_byname(struct mbox_client *cl,
 }
 EXPORT_SYMBOL_GPL(mbox_request_channel_byname);
 
+static void  mbox_txdone_init(struct mbox_controller *mbox, int *tx_method)
+{
+	if (mbox->txdone_irq)
+		*tx_method = TXDONE_BY_IRQ;
+	else if (mbox->txdone_poll)
+		*tx_method = TXDONE_BY_POLL;
+	else /* It has to be ACK then */
+		*tx_method = TXDONE_BY_ACK;
+}
+
 /**
  * mbox_free_channel - The client relinquishes control of a mailbox
  *			channel by this call.
@@ -478,8 +488,7 @@ void mbox_free_channel(struct mbox_chan *chan)
 	scoped_guard(spinlock_irqsave, &chan->lock) {
 		chan->cl = NULL;
 		chan->active_req = NULL;
-		if (chan->txdone_method == TXDONE_BY_ACK)
-			chan->txdone_method = TXDONE_BY_POLL;
+		mbox_txdone_init(chan->mbox, &chan->txdone_method);
 	}
 
 	module_put(chan->mbox->dev->driver->owner);
@@ -511,12 +520,7 @@ int mbox_controller_register(struct mbox_controller *mbox)
 	if (!mbox || !mbox->dev || !mbox->ops || !mbox->num_chans)
 		return -EINVAL;
 
-	if (mbox->txdone_irq)
-		txdone = TXDONE_BY_IRQ;
-	else if (mbox->txdone_poll)
-		txdone = TXDONE_BY_POLL;
-	else /* It has to be ACK then */
-		txdone = TXDONE_BY_ACK;
+	mbox_txdone_init(mbox, &txdone);
 
 	if (txdone == TXDONE_BY_POLL) {
 
-- 
2.51.0

